addon/plugins/article-structure-plugin/commands/wrap-structure-content.ts

```diff
@@ -71,15 +68,93 @@ const wrapStructureContent = (
   };
 };
 
-function findInsertionContainer(selection: Selection, nodeType: NodeType) {
+function findInsertionContainer(
+  selection: Selection,
+  nodeType: NodeType,
+  schema: Schema,
+) {
   const { $from } = selection;
   for (let currentDepth = $from.depth; currentDepth >= 0; currentDepth--) {
     const currentAncestor = $from.node(currentDepth);
     const pos = currentDepth > 0 ? $from.before(currentDepth) : -1;
+
+    // Simple like for like replacement from 0th index
     if (
       currentAncestor.canReplaceWith(0, currentAncestor.childCount, nodeType)
     ) {
-      return { node: currentAncestor, depth: currentDepth, pos };
+      return {
+        node: currentAncestor.content,
+        from: pos + 1,
+        to: pos + currentAncestor.nodeSize - 1,
+      };
+    }
+
+    const currentAncestorParent = $from.node(currentDepth - 1);
+    const currentAncestorIndexInParent = $from.index(currentDepth - 1);
+
+    if (!currentAncestorParent) {
+      return null;
+    }
+
+    // Sometimes we might not be able to replace from 0th index, but we can try to go one
+    // level up and attempt to replace `currentAncestor` as a child of its parent
+    if (
+      currentAncestorParent.canReplaceWith(
+        currentAncestorIndexInParent,
+        currentAncestorIndexInParent + 1,
+        nodeType,
+      )
+    ) {
+      return {
+        node: currentAncestor,
+        from: pos,
+        to: pos + currentAncestor.nodeSize - 1,
+      };
+    }
+
+    /**
+     * Special case when we reach `doc` node, but cannot replace the content.
+     * e.g. we reached doc, and it contains "document_title article article",
+     * we might not be able to wrap single `article`, depending on the `content` definiton,
+     * of the `doc`. Therefore we need to find sibling nodes of same type and try
+     * to wrap them all together
+     */
```

Comment:
What you wrap is everything *between* nodes of the same type, not only the same type.
This means if you have `article normal_paragraph article`, it will wrap it all.
If you have `normal_paragraph article article`, it will only wrap both articles.

This is *different* from when you "wrap" further inside (so you use the regular wrapping instead of this special case).
E.g. `Chapter [ normal_paragraph article article ]` (with paragraph and article inside the title). If you now set cursor on article and click "heading", it will also wrap `normal_paragraph`.


I'm not deciding what is the most logical, but it is inconsistent which isn't good :)